[inductor] Improve stability of scaled softmax

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124616
* __->__ #124119

This adds a pattern which replaces:
```python
   scale(x) - scale(x).amax(dim, keepdim=True)
```
with
```python
   scale(x - x.amax(dim, keepdim=True))
```
where `scale` can be either multiplication or division by a scalar,
or a tensor that is broadcast in the `dim` dimension.

We can find this pattern inside of the decomposed graph of:
```python
F.softmax(scale(x), dim=dim)
```

This has the effect of both reducing the chance of hitting the `fma`
issue and also means we avoid recomputing `scale(x)` inside and outside
the reduction which may be significant if we can remove an extra division.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

peterbell10 committed:
Update

[ghstack-poisoned]

Chillee commented:
Curious - do you think the right place to apply this is at the graph level or perhaps lower inside Inductor?

lezcano commented:
Note that the general pattern here would be `a*x - F(a*x)` = `a*(x-F(x))` for `F` linear, as the same thing may happen for any function `F` that's linear and that it happens to be the identity at some point (advanced indexing, diag + unsqueeze...). Detecting this may be possible if we look at the `auto_linear` tag in the derivatives file (if this is accessible). If it's not, we could always list the functions, but it may very well be overkill. If you think it is, just having this simpler pattern for starters looks reasonable to me.

peterbell10 commented:
> Curious - do you think the right place to apply this is at the graph level or perhaps lower inside Inductor?

I think this has to be done at the graph level as reductions are always realized, so the pattern would be split across two computed buffers. One buffer does `y = amax(scale(x))` and the other does `scale(x) - y`.

> Note that the general pattern here would be `a*x - F(a*x) = a*(x-F(x))` for `F` linear

While that pattern does make mathematical sense, I don't think it works here as `amax` isn't actually linear.

peterbell10 committed:
Update on "[inductor] Improve stability of scaled softmax"


This adds a pattern which replaces:
```python
   scale(x) - scale(x).amax(dim, keepdim=True)
```
with
```python
   scale(x - x.amax(dim, keepdim=True))
```
where `scale` can be either multiplication or division by a scalar,
or a tensor that is broadcast in the `dim` dimension.

We can find this pattern inside of the decomposed graph of:
```python
F.softmax(scale(x), dim=dim)
```

This has the effect of both reducing the chance of hitting the `fma`
issue and also means we avoid recomputing `scale(x)` inside and outside
the reduction which may be significant if we can remove an extra division.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

lezcano commented:
yeah, the more general pattern just asks for F to be [1-homogeneous](https://en.wikipedia.org/wiki/Homogeneous_function), but we don't have a good way of marking which ops are 1-homogeneous. (I realised of this after writing this but I don't currently have much access to internet lol).

We could write a small tool that uses OpInfos to figure out which ops are 1-homogeneous (or, in general, have any property that is stable under compositions (it's categorical in some sense)). Of course this is well outside this PR, but I may pitch this problem to someone, as I think it would be quite useful to write mathematically sound transformations.

lezcano commented on test/inductor/test_pattern_matcher.py:
```diff
@@ -1161,6 +1170,42 @@ def fn1(a, b):
+        def div_softmax(x, inv_scale):
+            return F.softmax(x / inv_scale, dim=0)
+
+        x = torch.randn(10, 10)
```
Should we use in this test very large numbers, like the ones that were showing NaNs in the example that motivated all this?

peterbell10 committed:
Use larger scale in test cases

[ghstack-poisoned]

peterbell10 committed:
Fix accidental revert

[ghstack-poisoned]

lezcano commented on torch/_inductor/fx_passes/joint_graph.py:
```diff
@@ -333,3 +344,119 @@ def pointless_view(match: Match, arg, size):
+#
+# which is more stable as we only compute the scaling once.
+#
+# NOTE: This pattern must come after fused attention matching!
```
Can we test this?

peterbell10 replied:
I think the existing fused attention tests are sufficient. If this pattern is applied first then the attention patterns wouldn't match any more.

lezcano reviewed (approved):
Nice! Just one question. What happens to the other PRs after this one? Are still relevant? Do we still want to disable fma at large?

peterbell10 commented:
> What happens to the other PRs after this one? Are still relevant? Do we still want to disable fma at large?

This is a tough question. On the stack disabling fma I'm seeing more accuracy failures after already fixing a handful. Given that the goal is ostensibly to improve accuracy, this isn't a great sign.

I think it might be worth exploring an fx graph level fma insertion. That fixes potential discrepancies between the same value being recomputed in different places, while still preserving the accuracy improvements from fma. However, there's still the issue that many of these operations aren't visible at the fx graph level. Once we reach inductor IR it's too late though as we don't know if an intermediate has multiple users.

peterbell10 commented:
@pytorchbot merge

peterbell10 committed:
Fix perf regression

[ghstack-poisoned]

lezcano commented on torch/_inductor/fx_passes/joint_graph.py:
```diff
@@ -333,3 +344,129 @@ def pointless_view(match: Match, arg, size):
+        if isinstance(other, (int, float)):
+            sign = 1 if other >= 0 else -1
+        else:
+            one = torch.scalar_tensor(1, dtype=inp.dtype, device=inp.device)
+            sign = torch.where(other >= 0, one, -one)
```
nit. you can even
```suggestion
            sign = torch.where(other >= 0, 1, -1)
```

peterbell10 replied:
I need these to be tensors so the where returns the correct dtype.

lezcano commented:
Why did the last commit fixes a perf regression?

peterbell10 commented:
> Why did the last commit fixes a perf regression?

Basically `where(other >= 0, inp, -inp)` was being saved for backwards, which resulted in additional buffers and writes, even if `other` was a scalar. On the other hand if I do `inp * where(other >= 0, one, -one)` then the multiplication can be recomputed in the backward pass from `inp`. Also changing `other.abs()` to `sign * other` allows better constant folding.

peterbell10 committed:
Update

[ghstack-poisoned]

peterbell10 committed:
Update

[ghstack-poisoned]

peterbell10 committed:
rebase

[ghstack-poisoned]

peterbell10 committed:
Update

[ghstack-poisoned]
